text/0016-internationalization.md

```diff
@@ -0,0 +1,117 @@
+    
+4.  Presently, OpenMRS uses Transifex ([https://www.transifex.com/openmrs/OpenMRS/dashboard/](https://www.transifex.com/openmrs/OpenMRS/dashboard/)) to manage translations. We will continue to use Transifex to support translations for the MF project.
+    
+5.  Possibly find ways to reuse OpenMRS module translations found in the `message.properties` file.
+
+7.  Update mechanism: Once translations are marked as complete, create a mechanism to publish the translated strings to a github repo. 
+    
+
+## Reason for decision
+
+- Most distributions are deployed to areas with no to intermittent connectivity, thus the translations module should be completely offline.
+
+- English is the lingua franca and spoken in most parts of the world.
+
+- Reasons for choosing Transifex:
+	 - Transifex is the most popular internationalization SaaS and offers tools that would make the translations workflow smoother.
+	 - The community is already familiar using transifex for translation services.
+
+
+## Alternatives
+- No translation service at all and default to English for all frontend modules.
+-  Use another service provider built specifically for frontend applications such as [react-i18next](https://github.com/i18next/react-i18next)
+- We could store the translations with each `esm` module instead of having a translations `esm` module.
+- Repo to handle translations service: we could have the translations handled by `openmrs-esm-api`
+
+
+## Common practices (not enforced)
+
+
+### Definitions
+
+Messages - OpenMRS historically used this word to refer to the content that would be replaced.
+
+  
+### Structure of Message File
+
+Here’s an example of how message.properties file looks like:
+
+[https://github.com/openmrs/openmrs-module-legacyui/blob/master/api/src/main/resources/messages.properties](https://github.com/openmrs/openmrs-module-legacyui/blob/master/api/src/main/resources/messages.properties)
+
+
+### Existing API
+
+Here is a [link](https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/messagesource/MessageSourceService.java) to the github repo for the interface to the java MessageSourceService in OpenMRS core. It has four methods:
+
+    public String getMessage(String s);
+    
+    public MutableMessageSource getActiveMessageSource();
+    
+    public  void  setActiveMessageSource(MutableMessageSource activeMessageSource);
+    
+    public  void  setMessageSources(Set<MutableMessageSource>  availableMessageSources);
+
+  
+
+### Proposed API
+
+1.  `setPreferredLocale()`: sets the preferred locale of the user, defaults to English if the user has no preferred locale.
+    
+2.  `getMessages(translation)` : takes argument `es-mx` and returns all the messages for that locale.
+    
+3.  `translate(messageVariableName, englishMessage, target-language)` : takes a message and translates it to the target language, if target language is not found, it returns the englishMessage
```

Comment:
@bmamlin my point is that the `targetLanguage` function is not going to be used because we'll be relying on the OpenMRS locale set in the user profile -- my original comment was questioning the need to allow any string to be translated to any language instead of just always translating to the locale specified in the user profile.

We have decisions to make about how to pull this off technically and allowing for any targetLanguage to be passed in would complicate it. The decisions are what to do at build time vs runtime, and if things are happening at runtime when do we download the translation strings. Note that the translate function here is synchronous, but that downloading extra translations is asynchronous. In order to make that possible, we have to know upfront which strings and language(s) to download before anybody calls the `translate` function.

My vote is for not allowing the targetLanguage to be specified here in the synchronous translate function.